File: python/categorized_items/backstage_pass.py
# -*- coding: utf-8 -*-
from categorized_item import CategorizedItem

class BackstagePass(CategorizedItem):
    def __init__(self, item):
        super().__init__(item)

    def update_sell_in(self):
        self.decrease_sell_in()

    def update_quality(self):
        if self.get_sell_in() >= 0:
            quality_increment = self._calculate_quality_increment()
            self.increase_quality(quality_increment)
        else:
            self.set_quality(0)

    def _calculate_quality_increment(self):
        quality_increment = 1
        if self.get_sell_in() < 10:
            quality_increment += 1
        if self.get_sell_in() < 5:
            quality_increment += 1
        return quality_increment

File: cpp/src/categorized_items/BackstagePass.cc
#include "BackstagePass.h"

BackstagePass::BackstagePass(Item &item) : CategorizedItem(item){}

void BackstagePass::updateQuality(){
    if (CategorizedItem::getSellIn() >= 0) {
        int qualityIncrease = BackstagePass::calculateQualityIncrease();
        CategorizedItem::increaseQualityBy(qualityIncrease);
    } else {
        BackstagePass::setQuality(0);
    }
}

int BackstagePass::calculateQualityIncrease(){
    int qualityIncrease = 1;
    if (BackstagePass::getSellIn() < 10){
        qualityIncrease++;
    }
    if (BackstagePass::getSellIn() < 5){
        qualityIncrease++;
    }
    return qualityIncrease;
}

void BackstagePass::updateSellIn(){
    CategorizedItem::decrementSellIn();
}